### ISSUE TITLE
Crash when increasing loan amount in construction mode

### ISSUE THREAD

**Issue by thetorpedodog** [OpenRCT2/OpenRCT2#8016]
**OS:** Ubuntu 18.04
**Version:** `0.2.1-develop-0~6915-201809201258-g2aa9765~ubuntu18.04.1`
**Commit/Build:** 2aa9765

Game crashed while I was building a coaster and had the finance window open and was borrowing money.

- [ ] Reproducible in RCT2 (vanilla)?
- [ ] Specific to multiplayer?

Not yet sure if it's reproducible. If I can generate a reliable repro I will do so.

**Dump file:** [Debian-style crash dump, hosted externally](https://pfish.zone/~thetorpedodog/_usr_bin_openrct2.1000.crash.gz)

**Save game:** [Autosaved game from several minutes before the crash](https://github.com/OpenRCT2/OpenRCT2/files/2408386/autosave_2018-09-22_23-51-57.sv6.gz)

**Comment by thetorpedodog** [comment]
EXCITING SLIGHTLY UPDATED SAVEGAME where just opening up the game (and maybe panning?) causes it to _immediately_ die for me:

[Crumbl.sv6.gz](https://github.com/OpenRCT2/OpenRCT2/files/2408403/Crumbl.sv6.gz)

**Comment by thetorpedodog** [comment]
I loaded the game, paused, and pincered the mechanic and set them down a few tiles away. They got a call and started responding to a breakdown on Whiplash, again. Odd.

**Comment by thetorpedodog** [comment]
Also, if I go to Whiplash and try to force a breakdown, it says:

![Can't force breakdown / Ride is already broken down](https://user-images.githubusercontent.com/6622003/46239731-8e879500-c36a-11e8-9d2a-346ccf8b0ea1.png)

**Comment by Gymnasiast** [comment]
@tombomp Exactly. And if I modify some code to prevent the assert from geting hit, the mechanic walks in the air at Whiplash's height. Pretty weird stuff.

**Comment by moonheart08** [comment]
Are we absolutely sure this save isn't being possessed by a demon?

**Comment by w-flo** [comment]
So this first part of my comment is about the strange "the ride is working fine, why is it getting repaired and why does it say it is already broken down when I want to force it to break down?" issue. I suspect it is because the ride is in the RIDE_LIFECYCLE_BREAKDOWN_PENDING lifecycle state, and at the same time it is in the RIDE_MECHANIC_STATUS_CALLING mechanic state. There are some checks that try to ensure this doesn't happen (I noticed that initiating the process to call a mechanic for an inspection is only triggered if there is no pending breakdown), but it might be possible like this:
- initially there is no breakdown and no pending breakdown
- ride inspection is due, so the CALLING mechanic state is activated to call a mechanic for inspection
- mechanic starts heading to ride for inspection
- now a pending breakdown is added, which is not visible to the player yet because the game waits for the broken car to arrive at the station
- mechanic gives up on heading to the ride for an inspection for whatever reason (picked up by player, maybe pathfinding issues or maybe it would be cancelled if another ride breaks down?)
- the ride finds out that the mechanic is no longer heading to the ride in ride_mechanic_status_update() even though it is in the RIDE_MECHANIC_STATUS_HEADING mechanic state, so it switches back to the RIDE_MECHANIC_STATUS_CALLING state
- now an idle mechanic is called again, but ride_call_closest_mechanic() sees the pending breakdown and calls the mechanic for repairing the ride instead of inspecting it.

So I believe this "mechanic wants to repair ride that appears to work fine" is not much of an issue really, mostly cosmetics. Maybe the "ride calls mechanic" state machine could be changed so it changes to UNDEFINED instead of CALLING if the current status is HEADING, but there is not actually a mechanic heading to the ride anymore _and_ there is a pending breakdown.

------------------------------------------

The second part of this comment is about the "why does the mechanic want to repair a non-existent car"? I suspect the root cause for this might be a missing "tailCar" for that corkscrew roller coaster. The train is actually supposed to have 1 more car than the number of cars specified in the user interface. The extra car is a very small car, at the end of the train, and has no seats. I can't see this car in this save, so maybe there is an issue with how that coaster train was created internally, or the car somehow disappeared after some time because of a bug. Maybe you @thetorpedodog still have an old savegame and could check if the car is there.

After closing that ride 2 times (so the train disappears) and then re-opening the ride, the train looks fine. And I believe the mechanic wanted to repair that tailCar because that one would have car index 5, but as @tombomp said, ride_get_broken_vehicle either crashes or returns nullptr (this is what happens for me) because that car index 5 does not exist.

Of course, the _actual_ root cause question would then be "why is there no tailCar"?

**Comment by jensj12** [comment]
Having experienced a similar crash a few years ago (and reported), I think the cause of the crash is indeed a discrepancy between the number of vehicles the ride thinks it has, and the actual number of vehicles on the ride.
Now the cause of that discrepancy can vary (it was a cheat + random vehicle reset a few years ago, now a missing tail car) and I’m not sure if there is or will ever be a legal way to do it (some folks do crazy thinks with the vehicles on a ride). I think there were some guards written as fix for my crash, but it may not have been enough.
